### PR TITLE
Update models to match alembic migrations

### PR DESCRIPTION
**Pull request by micahflee**
This PR tweaks the models so that when you autogenerate a migration with alembic, it does not need to create a new one.

For example, from the `main` branch:

```sh
$ rm instance/hushline.db
$ make migrate
$ make revision message="test"
. ./dev_env.sh && \
poetry run flask db upgrade
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 691dba936e64, Initial migration
INFO  [alembic.runtime.migration] Running upgrade 691dba936e64 -> ff59dfd3bdf6, add AuthenticationLog
INFO  [alembic.runtime.migration] Running upgrade ff59dfd3bdf6 -> 568f77aefcb4, add timecode to AuthenticationLog
INFO  [alembic.runtime.migration] Running upgrade 568f77aefcb4 -> 6e53eac9ea14, smtp encryption protocol
INFO  [alembic.runtime.migration] Running upgrade 6e53eac9ea14 -> 166a3402c391, add extra fields to user
. ./dev_env.sh && \
poetry run flask db revision -m "test" --autogenerate
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected removed index 'idx_authentication_logs_user_id_timestamp_successful' on 'authentication_logs'
INFO  [alembic.autogenerate.compare] Detected NULL on column 'message.content'
INFO  [alembic.autogenerate.compare] Detected NOT NULL on column 'users.password_hash'
INFO  [alembic.autogenerate.compare] Detected NOT NULL on column 'users.is_verified'
INFO  [alembic.autogenerate.compare] Detected NOT NULL on column 'users.is_admin'
INFO  [alembic.autogenerate.compare] Detected NOT NULL on column 'users.show_in_directory'
  Generating /home/user/code/scidsg/hushline/migrations/versions/a961a97b5f17_test.py ...  done
```

Here's the migration it created:

```python
"""test

Revision ID: a961a97b5f17
Revises: 166a3402c391
Create Date: 2024-08-29 08:42:38.998331

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = 'a961a97b5f17'
down_revision = '166a3402c391'
branch_labels = None
depends_on = None


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('authentication_logs', schema=None) as batch_op:
        batch_op.drop_index('idx_authentication_logs_user_id_timestamp_successful')

    with op.batch_alter_table('message', schema=None) as batch_op:
        batch_op.alter_column('content',
               existing_type=sa.TEXT(),
               nullable=True)

    with op.batch_alter_table('users', schema=None) as batch_op:
        batch_op.alter_column('password_hash',
               existing_type=sa.VARCHAR(length=512),
               nullable=False)
        batch_op.alter_column('is_verified',
               existing_type=sa.BOOLEAN(),
               nullable=False)
        batch_op.alter_column('is_admin',
               existing_type=sa.BOOLEAN(),
               nullable=False)
        batch_op.alter_column('show_in_directory',
               existing_type=sa.BOOLEAN(),
               nullable=False)

    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('users', schema=None) as batch_op:
        batch_op.alter_column('show_in_directory',
               existing_type=sa.BOOLEAN(),
               nullable=True)
        batch_op.alter_column('is_admin',
               existing_type=sa.BOOLEAN(),
               nullable=True)
        batch_op.alter_column('is_verified',
               existing_type=sa.BOOLEAN(),
               nullable=True)
        batch_op.alter_column('password_hash',
               existing_type=sa.VARCHAR(length=512),
               nullable=True)

    with op.batch_alter_table('message', schema=None) as batch_op:
        batch_op.alter_column('content',
               existing_type=sa.TEXT(),
               nullable=False)

    with op.batch_alter_table('authentication_logs', schema=None) as batch_op:
        batch_op.create_index('idx_authentication_logs_user_id_timestamp_successful', ['user_id', 'timestamp', 'successful'], unique=False)

    # ### end Alembic commands ###
```

However from this branch:

```sh
$ rm instance/hushline.db
$ make migrate
$ make revision message="test"
. ./dev_env.sh && \
poetry run flask db revision -m "test" --autogenerate
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.env] No changes in schema detected.
```

It no longer needs to autogenerate a migration.

### CODE REVIEW

**Review by glenn-sorrentino — APPROVED** [review]

File: hushline/model.py [hunk 1]
```diff
@@ -33,17 +34,17 @@ class User(Model):
     id: Mapped[int] = mapped_column(primary_key=True)
     primary_username: Mapped[str] = mapped_column(db.String(80), unique=True)
     display_name: Mapped[Optional[str]] = mapped_column(db.String(80))
-    _password_hash: Mapped[str] = mapped_column("password_hash", db.String(512))
+    _password_hash: Mapped[str] = mapped_column("password_hash", db.String(512), nullable=True)
```
Comment (jeremywmoore) [inline hushline/model.py:37]: I believe the proper way to indicate nullable using declarative table mapping is
```suggestion
    _password_hash: Mapped[Optional[str]] = mapped_column("password_hash", db.String(512))
```

File: hushline/model.py [hunk 2]
```diff
@@ -209,7 +219,7 @@ class SecondaryUsername(Model):
 class Message(Model):
     id: Mapped[int] = mapped_column(primary_key=True)
     _content: Mapped[Optional[str]] = mapped_column(
```
Comment (jeremywmoore) [inline hushline/model.py:221]: Removing the `Optional` should indicate that this column is not nullable.
```suggestion
    _content: Mapped[str] = mapped_column(
```

File: hushline/model.py [hunk 3]
```diff
@@ -33,17 +34,17 @@ class User(Model):
...
     _email: Mapped[Optional[str]] = mapped_column("email", db.String(255))
     _smtp_server: Mapped[Optional[str]] = mapped_column("smtp_server", db.String(255))
     smtp_port: Mapped[Optional[int]]
     _smtp_username: Mapped[Optional[str]] = mapped_column("smtp_username", db.String(255))
     _smtp_password: Mapped[Optional[str]] = mapped_column("smtp_password", db.String(255))
     _pgp_key: Mapped[Optional[str]] = mapped_column("pgp_key", db.Text)
-    is_verified: Mapped[bool] = mapped_column(default=False)
-    is_admin: Mapped[bool] = mapped_column(default=False)
-    show_in_directory: Mapped[bool] = mapped_column(default=False)
+    is_verified: Mapped[bool] = mapped_column(default=False, nullable=True)
```
Comment (jeremywmoore) [inline hushline/model.py:45]:
```suggestion
    is_verified: Mapped[Optional[bool]] = mapped_column(default=False)
```

File: hushline/model.py [hunk 4]
```diff
@@ -33,17 +34,17 @@ class User(Model):
...
+    is_admin: Mapped[bool] = mapped_column(default=False, nullable=True)
```
Comment (jeremywmoore) [inline hushline/model.py:46]:
```suggestion
    is_admin: Mapped[Optional[bool]] = mapped_column(default=False)
```

File: hushline/model.py [hunk 5]
```diff
@@ -33,17 +34,17 @@ class User(Model):
...
+    show_in_directory: Mapped[bool] = mapped_column(default=False, nullable=True)
```
Comment (jeremywmoore) [inline hushline/model.py:47]:
```suggestion
    show_in_directory: Mapped[Optional[bool]] = mapped_column(default=False)
```

**Review by jeremywmoore — CHANGES_REQUESTED** [review]

**Comment by jeremywmoore** [comment]
After experimenting a bit with my suggested changes, we might actually want to apply the autogenerated changes that this PR is attempting to squash. Boolean columns should probably have non-null value. We may also want to update the encrypt/decrypt methods to return `''`  instead of `None`.

**Comment by micahflee** [comment]
Okay this is ready for another review! It includes a migration that just makes those boolean fields required.

I briefly tried making the encrypt_field and decrypt_field functions always return a string, but that caused all sorts of problems. So instead I just made it so the Message.content setter stores a blank string if encrypt_fields returns None.

**Review by jeremywmoore — APPROVED** [review]

**Pull request merged**